text/0000-os-str-pattern.md

```diff
@@ -0,0 +1,527 @@
+
+    pub fn match_indices<'a, P>(&self, pat: P) -> impl Iterator<Item = (usize, &'a Self)>
+    where
+        P: Pattern<&'a Self>;
+
+    pub fn rmatch_indices<'a, P>(&self, pat: P) -> impl Iterator<Item = (usize, &'a Self)>
+    where
+        P: Pattern<&'a Self>,
+        P::Searcher: ReverseSearcher<&'a Self>;
+
+    // this is new
+    pub fn match_ranges<'a, P>(&'a self, pat: P) -> impl Iterator<Item = (Range<usize>, &'a Self)>
+    where
+        P: Pattern<&'a Self>;
+
+    // this is new
+    pub fn rmatch_ranges<'a, P>(&'a self, pat: P) -> impl Iterator<Item = (Range<usize>, &'a Self)>
+    where
+        P: Pattern<&'a Self>,
+        P::Searcher: ReverseSearcher<&'a Self>;
+
+    pub fn trim_matches<'a, P>(&'a self, pat: P) -> &'a Self
+    where
+        P: Pattern<&'a Self>,
+        P::Searcher: DoubleEndedSearcher<&'a Self>;
+
+    pub fn trim_left_matches<'a, P>(&'a self, pat: P) -> &'a Self
+    where
+        P: Pattern<&'a Self>;
+
+    pub fn trim_right_matches<'a, P>(&'a self, pat: P) -> &'a Self
+    where
+        P: Pattern<&'a Self>,
+        P::Searcher: ReverseSearcher<&'a Self>;
+
+    pub fn replace<'a, P>(&'a self, from: P, to: &'a Self) -> Self::Owned
+    where
+        P: Pattern<&'a Self>;
+
+    pub fn replacen<'a, P>(&'a self, from: P, to: &'a Self, count: usize) -> Self::Owned
+    where
+        P: Pattern<&'a Self>;
+}
+```
+
+We also allow slicing an `OsStr`.
+
+```rust
+impl Index<RangeFull> for OsStr { ... }
+impl Index<RangeFrom<usize>> for OsStr { ... }
+impl Index<RangeTo<usize>> for OsStr { ... }
+impl Index<Range<usize>> for OsStr { ... }
+```
+
+Example:
+
+```rust
+// (assume we are on Windows)
+
+let path = OsStr::new(r"C:\Users\Admin\😀\😁😂😃😄.txt");
+// can use starts_with, ends_with
+assert!(path.starts_with(OsStr::new(r"C:\")));
+assert!(path.ends_with(OsStr::new(".txt"));
+// can use rfind_range to get the range of substring
+let last_backslash = path.rfind_range(OsStr::new(r"\")).unwrap();
+assert_eq!(last_backslash, 16..17);
+// can perform slicing.
+let file_name = &path[last_backslash.end..];
+// can perform splitting, even if it results in invalid Unicode!
+let mut parts = file_name.split(&*OsString::from_wide(&[0xd83d]));
+assert_eq!(parts.next(), Some(OsStr::new("")));
+assert_eq!(parts.next(), Some(&*OsString::from_wide(&[0xde01])));
+assert_eq!(parts.next(), Some(&*OsString::from_wide(&[0xde02])));
+assert_eq!(parts.next(), Some(&*OsString::from_wide(&[0xde03])));
+assert_eq!(parts.next(), Some(&*OsString::from_wide(&[0xde04, 0x2e, 0x74, 0x78, 0x74])));
+assert_eq!(parts.next(), None);
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+It is trivial to apply the pattern API to `OsStr` on platforms where it is just an `[u8]`. The main
+difficulty is on Windows where it is an `[u16]` encoded as WTF-8. This RFC thus focuses on Windows.
+
+We will generalize the encoding of `OsStr` to specify these two capabilities:
```

Comment:
https://rust-lang.github.io/rfcs/0517-io-os-reform.html#string-handling does mention WTF-8 by name, so this RFC should probably mention the OMG-WTF-8 encoding by name (as a definition of the encoding, not just a sample implementation) even if the encoding remains an implementation detail.

The implementation PR should also make sure to update internal documentation to say that the representation of `OsStr` on Windows becomes OMG-WTF-8, with a link to the encoding’s full specification.

Reply:
Mentioned.